backpack/core/derivatives/crossentropyloss.py

```diff
@@ -64,49 +69,56 @@ def _sqrt_hessian_sampled(
         sqrt_mc_h = (probs_unsqueezed - classes) / sqrt(M)
 
         if module.reduction == "mean":
-            N = module.input0.shape[0]
-            sqrt_mc_h /= sqrt(N)
+            sqrt_mc_h /= sqrt(self._get_mean_normalization(module.input0))
 
+        sqrt_mc_h = self._ungroup_batch_and_additional(sqrt_mc_h, *rearrange_info)
         return sqrt_mc_h
 
-    def _sum_hessian(self, module, g_inp, g_out):
+    def _sum_hessian(
+        self, module: CrossEntropyLoss, g_inp: Tuple[Tensor], g_out: Tuple[Tensor]
+    ) -> Tensor:
         self._check_2nd_order_parameters(module)
 
         probs = self._get_probs(module)
-        sum_H = diag(probs.sum(0)) - einsum("bi,bj->ij", (probs, probs))
+        probs, *_ = self._merge_batch_and_additional(probs)
+        sum_H = diag(probs.sum(0)) - einsum("bi,bj->ij", probs, probs)
 
         if module.reduction == "mean":
-            N = module.input0.shape[0]
-            sum_H /= N
+            sum_H /= self._get_mean_normalization(module.input0)
 
         return sum_H
 
-    def _make_hessian_mat_prod(self, module, g_inp, g_out):
-        """Multiplication of the input Hessian with a matrix."""
+    def _make_hessian_mat_prod(
+        self, module: CrossEntropyLoss, g_inp: Tuple[Tensor], g_out: Tuple[Tensor]
+    ) -> Callable[[Tensor], Tensor]:
         self._check_2nd_order_parameters(module)
 
         probs = self._get_probs(module)
+        probs, *rearrange_info = self._merge_batch_and_additional(probs)
 
         def hessian_mat_prod(mat):
-            Hmat = einsum("bi,cbi->cbi", (probs, mat)) - einsum(
-                "bi,bj,cbj->cbi", (probs, probs, mat)
+            Hmat = einsum("bi,cbi->cbi", probs, mat) - einsum(
+                "bi,bj,cbj->cbi", probs, probs, mat
```

Comment:
`mat.dim()` is equal to `input.dim() + 1`. This `einsum` will fail unless `input.dim() == 2`. 

The problem for this PR is that this function is currently only tested by the old test suite (2d inputs to cross-entropy loss). The new test suite does not implement tests for this function. It would be good to add them.

Reply:
outdated